src/main/java/hudson/plugins/copyartifact/CopyArtifact.java

```diff
@@ -257,36 +316,46 @@ public boolean isFingerprintArtifacts() {
     }
 
     @Override
-    public boolean perform(AbstractBuild<?,?> build, Launcher launcher, BuildListener listener)
-            throws InterruptedException, IOException {
-        upgradeIfNecessary(build.getProject());
+    public void perform(@Nonnull Run<?, ?> build, @Nonnull FilePath workspace, @Nonnull Launcher launcher, @Nonnull TaskListener listener) throws InterruptedException, IOException {
+        if (build instanceof AbstractBuild) {
+            upgradeIfNecessary(((AbstractBuild)build).getProject());
+        }
+
+        EnvVars env = build.getEnvironment(listener);
+        if (build instanceof AbstractBuild) {
+            env.overrideAll(((AbstractBuild)build).getBuildVariables()); // Add in matrix axes..
+        }
+
         PrintStream console = listener.getLogger();
         String expandedProject = project, expandedFilter = filter;
         String expandedExcludes = getExcludes();
         try {
-            EnvVars env = build.getEnvironment(listener);
-            env.overrideAll(build.getBuildVariables()); // Add in matrix axes..
             expandedProject = env.expand(project);
```

Comment:
This call to `expand` is at best useless for a workflow, which should already have done any variable expansion it wanted before calling the constructor.

I think it would suffice to set `env` to be an empty `new EnvVars()` if not dealing with an `AbstractBuild`, so that calls to `expand` here and elsewhere just become the identity function. (Just double-check that this is the only thing `Selector.getBuild` uses `env` for.)